packages/eslint-plugin/src/rules/no-invalid-void-type.ts

```diff
@@ -174,6 +174,120 @@ export default createRule<[Options], MessageIds>({
       );
     }
 
+    function getMembers(
+      node:
+        | TSESTree.BlockStatement
+        | TSESTree.ClassBody
+        | TSESTree.Program
+        | TSESTree.TSModuleBlock,
+    ):
+      | TSESTree.ClassElement[]
+      | TSESTree.ProgramStatement[]
+      | TSESTree.Statement[] {
+      switch (node.type) {
+        case AST_NODE_TYPES.ClassBody:
+        case AST_NODE_TYPES.Program:
+        case AST_NODE_TYPES.TSModuleBlock:
+        case AST_NODE_TYPES.BlockStatement:
+          return node.body;
+      }
+    }
+
+    function getParentFunctionDeclarationNode(
+      node: TSESTree.Node,
+    ): TSESTree.FunctionDeclaration | TSESTree.MethodDefinition | null {
+      let current = node.parent;
+      while (current) {
+        if (current.type === AST_NODE_TYPES.FunctionDeclaration) {
+          return current;
+        }
+
+        if (
+          current.type === AST_NODE_TYPES.FunctionExpression &&
+          current.parent.type === AST_NODE_TYPES.MethodDefinition
+        ) {
+          return current.parent;
+        }
+
+        current = current.parent;
+      }
+
+      return null;
+    }
+
+    function getFunctionDeclarationName(
+      node:
+        | TSESTree.FunctionDeclaration
+        | TSESTree.MethodDefinition
+        | TSESTree.TSDeclareFunction,
+    ): string | symbol | undefined {
+      if (
+        node.type === AST_NODE_TYPES.FunctionDeclaration ||
+        node.type === AST_NODE_TYPES.TSDeclareFunction
+      ) {
+        // For a `FunctionDeclaration` or `TSDeclareFunction` this may be `null` if
+        // and only if the parent is an `ExportDefaultDeclaration`.
+        //
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return node.id!.name;
+      }
+
+      return getStaticMemberAccessValue(node, context);
+    }
+
+    function hasOverloadMethods(
```

Comment:
I think this may be useful as a shared utility function for different rules (might be better named `isOverloadImplementation`?). Re-using this function simplifies tackling https://github.com/typescript-eslint/typescript-eslint/issues/4586, for example.

Reply:
Adding that this attempts to emulate `checker.isImplementationOfOverload()` for rules which aren't type-aware.